### PR TITLE
[#67] Add build hook (4-2-stable)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #67 
Cherry-picked/Adapted from https://github.com/irods/irods_api_plugin_genquery2/pull/77

Left differences in a separate commit.

Still need to try this:
- [x] Centos 7
- [x] Ubuntu 18.04

### CODE REVIEW

**Review by korydraughn — COMMENTED** [review]
Looks ok.

I can't remember why we need spdlog for 4.2 though. I can look that up separately.

**Comment by alanking** [comment]
Hmm... I'm having some trouble building with this at the moment. Seems `charconv` header (https://en.cppreference.com/w/cpp/header/charconv) is missing, but it should be included with C++17, which is what 4.2.12 is using in irods/irods...

https://github.com/irods/irods_api_plugin_genquery2/blob/d4cc95676fc1f750329fd202eeb827595a91a642/CMakeLists.txt#L15
https://github.com/irods/irods/blob/2ed549ca7fe455aaa7755becc6c14b233dcbc0b4/CMakeLists.txt#L67-L68

**Comment by korydraughn** [comment]
Sounds like clang 6's support for c++17 doesn't have that implemented. Will have to tweak the code to make it work.

**Comment by alanking** [comment]
I've created an issue for the build failure here: https://github.com/irods/irods_api_plugin_genquery2/issues/79

**Comment by korydraughn** [comment]
Ok. I noticed you didn't assign yourself to the issue. Feel free to take care of it if you're up to it.

**Comment by alanking** [comment]
Added fix for #79. I haven't tested the change yet, but have confirmed that the plugin builds for centos 7 and ubuntu 18 using the build hook.

**Comment by alanking** [comment]
Squashed and marked ready for review. Still need to test, tho

File: parser/dsl/lexer.l [hunk 1]
```diff
@@ -41,14 +41,19 @@
...
-                                throw yy::parser::syntax_error{loc, fmt::format("invalid hex value: [{}]", YYText())};
-                            }
-                            else if (err.ec == std::errc::result_out_of_range) {
-                                throw yy::parser::syntax_error{loc, fmt::format("hex value cannot be represented: [{}]", YYText())};
+                            // Interpret the hex value using std::sscanf and assign to value. If ret is less than 1,
+                            // this means that the single receiving argument (&value) was not successfully assigned.
+                            std::uint8_t value;
+                            if (const int ret = std::sscanf(buffer, "%2x", &value); ret < 1) {
+                                throw yy::parser::syntax_error{
+                                    loc, fmt::format("failed to assign hex value: [{}]", YYText())};
```
Comment (korydraughn) [inline parser/dsl/lexer.l:56]: Let's restore the error message to what the previous wording to keep both branches in sync.

File: parser/dsl/lexer.l [hunk 2]
```diff
@@ -41,14 +41,19 @@
...
 
-                            if (err.ec == std::errc::invalid_argument) {
-                                throw yy::parser::syntax_error{loc, fmt::format("invalid hex value: [{}]", YYText())};
-                            }
-                            else if (err.ec == std::errc::result_out_of_range) {
-                                throw yy::parser::syntax_error{loc, fmt::format("hex value cannot be represented: [{}]", YYText())};
+                            // Interpret the hex value using std::sscanf and assign to value. If ret is less than 1,
+                            // this means that the single receiving argument (&value) was not successfully assigned.
+                            std::uint8_t value;
+                            if (const int ret = std::sscanf(buffer, "%2x", &value); ret < 1) {
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply parser/dsl/lexer.l:54]: `std::sscanf` feels unsafe every time I see, but I also can't say I've used it enough to make any claims against it.

IMO, `strtoul` and `stoul` look safer and faster given their interfaces.

[the remark this answers is not in the document]
Comment (alanking) [reply parser/dsl/lexer.l:54]: There are two reasons I picked `sscanf`:

1. `std::from_chars` specifically mentions `std::sscanf` as an alternative here: https://en.cppreference.com/w/cpp/utility/from_chars.
2. I didn't know that `strtoul`/`stoul` could interpret hex values. I can try that if it feels better.

[the remark this answers is not in the document]
Comment (korydraughn) [reply parser/dsl/lexer.l:54]: I didn't know `from_chars` referenced `sscanf`.

I'm mostly not familiar with it enough to state whether it's a good tool or not. I feel the others are probably better/easier, but we should investigate that.

We can do that for this PR or create an issue and investigate it later. Up to you.

File: parser/dsl/lexer.l (hunk 1 above)
Comment (korydraughn) [reply parser/dsl/lexer.l:56]: I see the issue. There's only one error case now.

It's fine as is.

File: parser/dsl/lexer.l (hunk 2 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply parser/dsl/lexer.l:54]: The unsigned versions seem fine too. I wonder which versions are more appropriate for this?

[the remark this answers is not in the document]
Comment (alanking) [reply parser/dsl/lexer.l]: I used `std::stoul` but switching to any of the others should be trivial. Once again, still need to test.

File: parser/dsl/lexer.l [hunk 3]
```diff
@@ -41,17 +40,27 @@
...
 
-                            if (err.ec == std::errc::invalid_argument) {
+                                // Given that this is a base-16 value with 2 digits, its maximum value is \xFF, or 255.
+                                // The maximum value of std::uint8_t is UINT8_MAX, or 255. The value is guaranteed to
+                                // fit into the type despite a loss in precision (unsigned long -> std::uint8_t).
+                                std::uint8_t value = std::stoul(buffer, nullptr, 16);
+
+                                string_literal.push_back(value);
+                            }
+                            catch (const std::invalid_argument& e) {
```
Comment (alanking) [inline parser/dsl/lexer.l:58]: Whoops, need to get rid of the named variables. Not using `e`.

File: parser/dsl/lexer.l [hunk 4]
```diff
@@ -41,17 +40,27 @@
...
+                                constexpr auto num_chars = 2;
+                                constexpr auto buffer_len = num_chars + 1;
+                                char buffer[buffer_len]{};
+                                std::memcpy(buffer, YYText() + 2, num_chars);
 
-                            if (err.ec == std::errc::invalid_argument) {
+                                // Given that this is a base-16 value with 2 digits, its maximum value is \xFF, or 255.
+                                // The maximum value of std::uint8_t is UINT8_MAX, or 255. The value is guaranteed to
+                                // fit into the type despite a loss in precision (unsigned long -> std::uint8_t).
+                                std::uint8_t value = std::stoul(buffer, nullptr, 16);
```
Comment (korydraughn) [inline parser/dsl/lexer.l:54]: Could be `const`.

**Review by korydraughn — COMMENTED** [review]
I think this is a winner.

File: parser/dsl/lexer.l (hunk 3 above)
Comment (alanking) [reply parser/dsl/lexer.l]: Done

File: parser/dsl/lexer.l (hunk 4 above)
Comment (alanking) [reply parser/dsl/lexer.l]: Done

File: parser/dsl/lexer.l (hunk 2 above)
[the remark this answers is not in the document]
Comment (alanking) [reply parser/dsl/lexer.l]: Resolving. We can open a new conversation if we want to go with another member of that family of functions.

**Comment by alanking** [comment]
Hey, I figured out how to test this thing!
```bash
$ ils -l
/tempZone/home/rods:
  rods              0 demoResc          283 2023-11-10.17:08 & Hello, GenQuery2!
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x21'"
[["10017"]]
```

**Comment by korydraughn** [comment]
It's so cool!

**Comment by korydraughn** [comment]
Try the following:
- A query with characters immediately following the hex byte (`\x21alanking`, `\x21a`)
- A query with an invalid hex byte (out of range, invalid character e.g. `\xXY`)

**Comment by alanking** [comment]
```bash
$ ils -l
/tempZone/home/rods:
  rods              0 demoResc          283 2023-11-10.17:08 & Hello, GenQuery2!
  rods              0 demoResc          283 2023-11-10.17:16 & Hello, GenQuery2!alanking
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x21alanking'"
[["10018"]]
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\xXY'"
[]
```

I did find one quirk. Not sure if this is expected...
```bash
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\xff'"
error: -167000
```

In the log:
```
Nov 10 17:10:37 pid:464 remote addresses: 172.29.0.2, 172.29.0.3 ERROR: Caught database exception while executing query: /externals/nanodbc2.13.0-1_src/nanodbc/nanodbc/nanodbc.cpp:1686: 2202: ERROR: invalid byte sequence for encoding "UTF8": 0xff;
Error while executing the query 
```

And here's the boundary:
```bash
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x80'"
error: -167000
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x79'"
[]
```

0x80 == 128. Doesn't seem random. But... I'm also not convinced this is my change. :)

**Comment by korydraughn** [comment]
Some of those are definitely surprising.

I'm going to try those same cases against main and let you know if anything is different.

**Comment by korydraughn** [comment]
Here's what I got.

```
$ ils -l
/tempZone/home/rods:
  rods              0 demoResc            0 2023-08-10.12:17 & []{}()#!@$^&,;%^.txt
  rods              0 demoResc            0 2023-08-10.12:24 & []{}()#\!@$^&,;'%^.txt
  rods              0 demoResc            0 2023-08-10.12:21 & \[\]\{\}\(\)\#\!\@$^\&\,\;\%\^\.txt
  rods              0 demoResc            0 2023-11-10.12:27 & Hello, Genquery2!
  rods              0 demoResc            0 2023-11-10.12:28 & Hello, GenQuery2!
  rods              0 demoResc            0 2023-11-10.12:29 & Hello, GenQuery2!alanking
  C- /tempZone/home/rods/irods431_pkgs  
  C- /tempZone/home/rods/tiering_ub18_cen7  
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x21'"
[["85097"]]
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x21alanking'"
[["85098"]]
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\xXY'"
[]
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\xff'"
error: -130000
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x80'"
error: -130000
$ iquery "select DATA_ID where DATA_NAME = 'Hello, GenQuery2\x79'"
[]
```

For `\xff`, the log contains ...

```json
{
  "log_category": "genquery2",
  "log_level": "error",
  "log_message": "[json.exception.type_error.316] invalid UTF-8 byte at index 34: 0xFF",
  "request_api_name": "",
  "request_api_number": 1000001,
  "request_api_version": "d",
  "request_client_user": "rods",
  "request_host": "[REDACTED]",
  "request_proxy_user": "rods",
  "request_release_version": "rods4.3.1",
  "server_host": "[REDACTED]",
  "server_pid": 4019747,
  "server_timestamp": "2023-11-10T17:32:12.359Z",
  "server_type": "agent",
  "server_zone": "tempZone"
}
```

I think the behavior of `\xff` may be okay for now. We'd need to investigate whether that byte is special in the context of UTF-8.

**Comment by korydraughn** [comment]
We are getting different error codes for some reason though.

**Comment by alanking** [comment]
Some thoughts...
 - There are differences in the externals versions (`json`, `nanodbc`, `clang`, at least)
 - The logger in the main branch uses `json`; not so in 4-2-stable
 - Whatever the difference is would probably otherwise be found in this change, I would think: https://github.com/irods/irods_api_plugin_genquery2/commit/bc78f1601055dc3c27b5e954899e8a082653fe53

The only difference between this change and the same section in the main branch is the machinery leading up to the `push_back`

**Comment by korydraughn** [comment]
Good point. Let's get this merged.

If you're happy with the changes, squash and pound it.

**Comment by alanking** [comment]
Done.

Created https://github.com/irods/irods_api_plugin_genquery2/issues/80 (nice) for the errors around 0x80.

**Comment by korydraughn** [comment]
Merging.

**Pull request merged**